File: ds/linked_list.py
class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = ListNode(-1)
        self.tail = self.head

    #O(1)
    def insertEnd(self, val):
        self.tail.next = ListNode(val)
        self.tail = self.tail.next

    #O(n)
    def removeEnd(self):
        curr = self.head.next
        while curr.next.next:
            curr = curr.next
        curr.next = None
        self.tail = curr

    #O(1)
    def removeFront(self):
        self.head.next = self.head.next.next

    def removeIndex(self, index):
        i = 0
        curr = self.head
        while i < index and curr:
            i += 1
            curr = curr.next

        if curr and curr.next:
            if curr.next == self.tail:
                self.tail = curr
            curr.next = curr.next.next

    #O(n)
    def print(self):
        curr = self.head.next
        while curr:
            print(curr.val)
            curr = curr.next

# test cases for linked list with assertion
def test_linked_list():
    ll = LinkedList()
    ll.insertEnd(1)
    ll.insertEnd(2)
    ll.insertEnd(3)
    ll.insertEnd(4)
    ll.insertEnd(5)

    ll.removeEnd()
    ll.removeIndex(2)
    ll.removeFront()

    assert ll.head.next.val == 2
    assert ll.tail.val == 4

    print('All test cases pass')

test_linked_list()

File: ds/linked_list.js
class ListNode {
    constructor(val){
        this.val = val;
        this.next = null;
    }
}

class LinkedList {
    constructor(){
        this.head = new ListNode(-1);
        this.tail = this.head;
    }

    insertEnd(val){
        let newNode = new ListNode(val);
        this.tail.next = newNode;
        this.tail = this.tail.next;
    }

    removeEnd(){
        let curr = this.head.next;
        while(curr.next.next){
            curr = curr.next;
        }
        curr.next = curr.next.next;
        this.tail = curr;
    }

    removeFront(){
        this.head.next = this.head.next.next;
    }

    removeIndex(index){
        let i = 0;
        let curr = this.head;
        while(i < index && curr){
            i++;
            curr = curr.next;
        }
        if(curr && curr.next){
            if(curr.next == this.tail){
                this.tail = curr;
            }
            curr.next = curr.next.next;
        }
    }
}

function TestLinkedList(){
    const assert = require('assert');
    let linkedList = new LinkedList();
    linkedList.insertEnd(1);
    linkedList.insertEnd(2);
    linkedList.insertEnd(3);
    linkedList.insertEnd(4);
    linkedList.insertEnd(5);

    linkedList.removeEnd();
    assert(linkedList.tail.val === 4);

    linkedList.removeFront();
    assert(linkedList.head.next.val === 2);

    linkedList.removeIndex(1);
    assert(linkedList.head.next.next.val === 4);

    console.log('All test cases pass');
}

TestLinkedList();